controllers/nstemplateset/cluster_resources.go

```diff
@@ -206,3 +151,99 @@ func (r *clusterResourcesManager) delete(ctx context.Context, nsTmplSet *toolcha
+// A new instance can be obtained using the newObjectApplier() function.
+//
+// It takes in the current state of the NSTemplateSet and then processes one new object to apply
+// to the cluster at a time, updating the internal bookkeeping to be able to Cleanup() the old objects
+// after all the new objects have been applied.
+type objectApplier struct {
+	r                   *clusterResourcesManager
+	statusChanged       bool
+	firstDeployment     bool
+	failureStatusReason statusUpdater
+	nstt                *toolchainv1alpha1.NSTemplateSet
+	currentObjects      []runtimeclient.Object
+	newTierTemplate     *tierTemplate
+}
+
+func newObjectApplier(r *clusterResourcesManager, nstt *toolchainv1alpha1.NSTemplateSet, currentObjects []runtimeclient.Object, newTierTemplate *tierTemplate) *objectApplier {
+	firstDeployment := nstt.Status.ClusterResources == nil || nstt.Status.ClusterResources.TemplateRef == ""
+	var failureStatusReason statusUpdater
+
+	if firstDeployment {
+		failureStatusReason = r.setStatusClusterResourcesProvisionFailed
+	} else {
+		failureStatusReason = r.setStatusUpdateFailed
+	}
+
+	return &objectApplier{
+		r:                   r,
+		statusChanged:       false,
+		firstDeployment:     firstDeployment,
+		failureStatusReason: failureStatusReason,
+		nstt:                nstt,
+		currentObjects:      currentObjects,
+		newTierTemplate:     newTierTemplate,
+	}
+}
+
+func (oa *objectApplier) changeStatusIfNeeded(ctx context.Context) error {
+	if !oa.statusChanged {
+		var err error
+		if oa.firstDeployment {
+			err = oa.r.setStatusProvisioningIfNotUpdating(ctx, oa.nstt)
+		} else {
+			err = oa.r.setStatusUpdatingIfNotProvisioning(ctx, oa.nstt)
+		}
+		if err != nil {
+			return err
+		}
+		oa.statusChanged = true
+	}
+	return nil
+}
+
+func (oa *objectApplier) Apply(ctx context.Context, obj runtimeclient.Object) error {
+	if !shouldCreate(obj, oa.nstt) {
+		return nil
+	}
+
+	// by removing the new objects from the current objects, we are going to be left with the objects
+	// that should be removed from the cluster after we're done with this loop
+	for oldIdx, curObj := range oa.currentObjects {
+		if commonclient.SameGVKandName(curObj, obj) {
+			oa.currentObjects = slices.Delete(oa.currentObjects, oldIdx, oldIdx+1)
+			break
+		}
+	}
+
+	if err := oa.changeStatusIfNeeded(ctx); err != nil {
+		return err
+	}
+
+	// create or update the resource
+	if _, err := oa.r.apply(ctx, oa.nstt, oa.newTierTemplate, obj); err != nil {
+		err := fmt.Errorf("failed to apply changes to the cluster resource %s, %s: %w", obj.GetName(), obj.GetObjectKind().GroupVersionKind().String(), err)
+		return oa.r.wrapErrorWithStatusUpdate(ctx, oa.nstt, oa.failureStatusReason, err, "failure while syncing cluster resources")
+	}
+
+	return nil
+}
+
+func (oa *objectApplier) Cleanup(ctx context.Context) error {
+	// what we're left with here is the list of currently existing objects that are no longer present in the template.
+	// we need to delete them
```

Comment:
or maybe just move this down "in front of" the `deleteObsoleteObjects` call and extend why it is so, thanks